File: src/main/java/org/warp/picalculator/math/rules/NumberRule7.java
package org.warp.picalculator.math.rules;

import it.unimi.dsi.fastutil.objects.ObjectArrayList;

import org.warp.picalculator.Error;
import org.warp.picalculator.math.MathContext;
import org.warp.picalculator.math.Function;
import org.warp.picalculator.math.functions.Multiplication;
import org.warp.picalculator.math.functions.Number;
import org.warp.picalculator.math.functions.Sum;

/**
 * Number rule<br>
 * <b>a + a = 2a</b>
 * 
 * @author Andrea Cavalli
 *
 */
public class NumberRule7 {

	public static boolean compare(Sum f) {
		return f.getParameter1().equals(f.getParameter2());
	}

	public static ObjectArrayList<Function> execute(Sum f) throws Error {
		final MathContext root = f.getMathContext();
		final ObjectArrayList<Function> result = new ObjectArrayList<>();
		final Multiplication mult = new Multiplication(root, new Number(root, 2), f.getParameter1());
		result.add(mult);
		return result;
	}

}

File: src/main/resources/rules/NumberRule7.js
//Imports


var Error = org.warp.picalculator.Error;
var Function = org.warp.picalculator.math.Function;
var MathContext = org.warp.picalculator.math.MathContext;
var Multiplication = org.warp.picalculator.math.functions.Multiplication;
var Number = org.warp.picalculator.math.functions.Number;
var Sum = org.warp.picalculator.math.functions.Sum;

var ObjectArrayList = Java.type("it.unimi.dsi.fastutil.objects.ObjectArrayList");

/**
 * Number rule
 * a + a = 2a
 * 
 * @author Andrea Cavalli
 *
 */
var rule = {
	// Rule name
	getRuleName: function() {
		return "NumberRule7";
	},
	// Rule type
	getRuleType: function() {
		return RuleType.EXPANSION;
	},
	/* Rule function
	   Returns:
	     - null if it's not executable on the function "f"
	     - An ObjectArrayList<Function> if it did something
	*/

	execute: function(f) {
		var isExecutable;
		if (ScriptUtils.instanceOf(f, Sum.class)) {
			isExecutable = f.getParameter1().equals(f.getParameter2());
		}
		
		if (isExecutable) {
			var root = f.getMathContext();
			var result = new ObjectArrayList();
			var mult = new Multiplication(root, new Number(root, 2), f.getParameter1());
			result.add(mult);
			return result;
		} else {
			return null;
		}
	}
}

//Add this rule to the list of rules
RulesManager.addRule(engine.getInterface(rule, Rule.class));